Fix #476 and #364: Improve contacts form UI

seav opened the pull request:
- Rework contacts form HTML and CSS to better present errors
- Update JavaScript to correctly maintain state of contacts form

Aside from other developers reviewing the code, @clash99 should also review if the UI/UX is as expected.

**Note:** #451 is not yet resolved because this requires uploading icon files to S3 first.

oliverroick commented on cadasta/organization/forms.py:
```diff
@@ -48,17 +48,26 @@ class ContactsForm(forms.Form):
                     '<span aria-hidden="true">&times;</span></a></td>')
+        html = ('' if self.errors else '<tr>') + html + closeBtn + '</tr>\n'
+
+        error_types = ''
```
There is now custom rendering of error classes here; you should add a test to check whether these are rendered as expected. You can add the test [below here](https://github.com/Cadasta/cadasta-platform/blob/c36daebdd48b8796a47a173f20958590b4f73792/cadasta/organization/tests/test_forms.py#L484).

seav replied:
I understand. I will add four tests for the different types of errors that can be encountered.

oliverroick commented on cadasta/organization/tests/test_forms.py:
```diff
@@ -581,16 +581,33 @@ def test_validate_valid_form_with_tel(self):
+        }
+        form = forms.ContactsForm(data=data, prefix='contacts')
+        assert form.is_valid() is False
+        assert form.errors['name']
```
Here (and in the following line) you should test whether the `name` key has the correct error message assigned. I.e. `assert form.errors['name'] == 'Please provide a name.' It's better to be specific; if, for instance, I will change the code later on that accidently sets name to 'You owe me 10 billion Dollars' this assertion will still pass.

seav replied:
I do not actually use the error messages in the `name` and `email` keys. The error messages I use are all stuffed into the `__all__` key so that we can have the error message that spans all table columns. These error messages will be tested with the 4 tests that I plan to create in the previous comment.

oliverroick commented:
This branch seems to be broken. I get a `JSONDecodeError` on all pages. (Even after I destroyed and reprovisioned the VM).

According to the error page, there seems to be something wrong with the source map. 

![screen shot 2016-07-25 at 10 46 11](https://cloud.githubusercontent.com/assets/159510/17095512/0b142202-5255-11e6-8da3-4c06af8df583.png)

![screen shot 2016-07-25 at 10 46 50](https://cloud.githubusercontent.com/assets/159510/17095524/1cd17e86-5255-11e6-9dcd-695f12ecaa19.png)

ian-ross commented:
The source map is my fault.  There was a conflict when rebasing.  Let me sort it out now...

oliverroick commented:
I have resolved the merge conflict. You can ignore the last to comments from Ian and me. Sorry for the trouble.

oliverroick commented:
One more issue, though: New form fields are sometimes marked as fields that have an error. You reproduce it by following:

1. Create a new organization or project, add an invalid contact with the first row in the table and submit the form.
2. You should then see the errors correctly. 
3. Remove the first row (the one with the error) from the table. 
4. Add a new contact. The rows that had an error before still have the red border.

seav commented:
Ah, yes. I forgot to clear out the error classes when cloning a new row. I'll update the JS.

seav commented:
Modifications are done:
- Added 4 tests to check the `as_table()` output for error cases
- Modified contacts.js to remove error classes in new contact rows

oliverroick commented:
Ok, this is good to go from my end. @clash99 can you please check if the UI/UX is all ok?

clash99 commented:
This looks ready to go.  Thanks everyone!
